<a href="https://colab.research.google.com/github/rohansiddam/Python-Journey/blob/main/014%20-%20Lesson%2014%20(Hunting%20Exoplanets%20In%20Space%20-%20Pandas%20DataFrame).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lesson 14: Hunting Exoplanets In Space - Pandas DataFrame

### Teacher-Student Activities

In the last class, we learnt about the Pandas series. Now in this lesson, we will learn about Pandas DataFrame which is a collection of Pandas series. In other words, a Pandas DataFrame is a two-dimensional array.

In the process of learning Pandas DataFrames, we will learn how NASA finds the exoplanets in the universe. There are deep Physical and mathematical theories on exploring exoplanets in the space. But we will not go through all of that. Right now, we just need to understand the basic principle behind these theories to be able to learn how the exoplanets are found.

---



### Finding Exoplanets Principle

Imagine that you are in your room during the day time with the window curtains open. The room probably would be well-lit. Now, imagine that you close the curtains of the window and block the sunlight from entering the room. In this situation, the room would be darker and the visibility would be low.

So, whenever the curtains are open, the brightness of the light would be higher whereas when the curtains are closed, the brightness would be lower. We can measure the brightness of the light using a spectroscope.

The same principle is applied in searching for an exoplanet. There are billions of galaxies in the universe. These galaxies have millions of stars. One such galaxy is the Milky-way galaxy in which our solar system exists. The solar system has a star called Sun which has its own light. In astronomy, a star is a heavenly body which has its own light. There are 8 planets in our solar system orbiting around the Sun. Similar to this, in some other galaxy there would be a star and probably a planet would be revolving around that star.

Long back, NASA placed a telescope called Kepler telescope in the space. This telescope is used to measure the brightness of the stars in the far-distant galaxies.


<img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/kepler-exoplanets-dataset/kepler-space-telescope.jpg' width="800">

*Image credits: http://kepler.nasa.gov/Mission/discoveries/discoverygraphics/*

Whenever a planet, while orbiting its star, comes in between the telescope and the star, the brightness of the star recorded by the telescope is lower whereas when the planet goes behind the star, the brightness of the light recorded by the telescope is higher.

This method of detecting exoplanets in far-distant galaxies through the brightness of the light emitted by a star is called the **Transit Method**. You can read about it by clicking on the link provided in the **Activities** section under the title **How Do Astronomers Find Exoplanets?**

Essentially, if you plot the brightness on the vertical axis and the time on the horizontal axis, then you will see that the brightness of the star recorded by the telescope increases and decreases periodically. Thus, in the graph, you will notice a wave-like pattern. This indicates that the star definitely has at least one planet.

<img src = 'https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/kepler-exoplanets-dataset/transit-method-gif.gif' width='800'>

*Image credits: https://imgur.com/r/SpaceGifs/43f17Ke*

The image below shows some of the exoplanets (Kepler 4b to Kepler 8b) discovered by the Kepler space telescope. You can see the brightness level radiated by the star for each planet. The Flux values on the vertical axis represent the brightness level of the star.

<img src = 'https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/kepler-exoplanets-dataset/transit-method.jpg' width='800'>

*Image credits: http://kepler.nasa.gov/Mission/discoveries/discoverygraphics/*

As you can see in the image above, the bigger the planet (Kepler 6b), deeper the dip in the brightness level. And, the longer the orbital period of a planet, broader is the width of the dip (Kepler 7b). Kepler 7b has the greatest orbital period of 4.9 days among these 5 planets.

So, this is how NASA finds a planet beyond our solar system. Now, let's use Kepler space telescope dataset to create a Pandas DataFrame in order to find out which stars beyond our solar system have a planet.

---

#### Activity 1: Loading A CSV File^^

Generally, we store data in different files such as text (`txt` format) file, comma-separated value (`csv` format) file, tab-separated value (`tsv` format) file etc. We can read the contents of these files through Python.

A comma-separated value (`csv`) file is used most commonly to store data.

To load or read the contents of a `csv` file, we can use the `read_csv()` function in Pandas. The data is read in the form of a two-dimensional array called a **Pandas DataFrame**.

As an input to the `read_csv()` function, we need to provide the full-location of the `csv` file that we wish to read in the string format. The file that we wish to read is stored in cloud storage.

Here's the link to the file.

https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/kepler-exoplanets-dataset/exoTrain.csv

In the above link, `exoTrain.csv` is the name of the `csv` file. We will create Pandas DataFrame using this file.

In [ ]:
# Teacher Action: Read a 'csv' file using the 'read_csv()' function. Also, display the first 5 rows of the DataFrame using the 'head()' function.
# First of all we have to import the Pandas module with pd as an alias (or nickname).
import pandas as pd
csv_file = "https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/kepler-exoplanets-dataset/exoTrain.csv"
exoTrain = pd.read_csv(csv_file)
exoTrain.head()

LABEL   FLUX.1   FLUX.2   FLUX.3   FLUX.4   FLUX.5   FLUX.6  FLUX.7  \
0      2    93.85    83.81    20.10   -26.98   -39.56  -124.71 -135.18   
1      2   -38.88   -33.83   -58.54   -40.09   -79.31   -72.81  -86.55   
2      2   532.64   535.92   513.73   496.92   456.45   466.00  464.50   
3      2   326.52   347.39   302.35   298.13   317.74   312.70  322.33   
4      2 -1107.21 -1112.59 -1118.95 -1095.10 -1057.55 -1034.48 -998.34   

    FLUX.8  FLUX.9  ...  FLUX.3188  FLUX.3189  FLUX.3190  FLUX.3191  \
0   -96.27  -79.89  ...     -78.07    -102.15    -102.15      25.13   
1   -85.33  -83.97  ...      -3.28     -32.21     -32.21     -24.89   
2   486.39  436.56  ...     -71.69      13.31      13.31     -29.89   
3   311.31  312.42  ...       5.71      -3.73      -3.73      30.05   
4 -1022.71 -989.57  ...    -594.37    -401.66    -401.66    -357.24   

   FLUX.3192  FLUX.3193  FLUX.3194  FLUX.3195  FLUX.3196  FLUX.3197  
0      48.57      92.54      39.32      61.42       5.08     -39.54  
1      -4.86       0.76     -11.70       6.46      16.00      19.93  
2     -20.88       5.06     -11.80     -28.91     -70.02     -96.67  
3      20.03     -12.67      -8.77     -17.31     -17.35      13.98  
4    -443.76    -438.54    -399.71    -384.65    -411.79    -510.54  

[5 rows x 3198 columns]

As you can see, we have created a Pandas DataFrame for the `exoTrain.csv` file and stored it in the `exo_train_df` variable.

Now, you create a DataFrame for the `exoTest.csv` file and store it in a variable called `exo_test_df`. Here is the link to the file.

https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/kepler-exoplanets-dataset/exoTest.csv

In [ ]:
# Student Action: Read the 'exoTest.csv' file and display its first 5 rows using the 'head()' function.
csv_file2 = "https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/kepler-exoplanets-dataset/exoTest.csv"
exoTest = pd.read_csv(csv_file2)
exoTest.head()

LABEL   FLUX.1   FLUX.2   FLUX.3   FLUX.4   FLUX.5   FLUX.6   FLUX.7  \
0      2   119.88   100.21    86.46    48.68    46.12    39.39    18.57   
1      2  5736.59  5699.98  5717.16  5692.73  5663.83  5631.16  5626.39   
2      2   844.48   817.49   770.07   675.01   605.52   499.45   440.77   
3      2  -826.00  -827.31  -846.12  -836.03  -745.50  -784.69  -791.22   
4      2   -39.57   -15.88    -9.16    -6.37   -16.13   -24.05    -0.90   

    FLUX.8   FLUX.9  ...  FLUX.3188  FLUX.3189  FLUX.3190  FLUX.3191  \
0     6.98     6.63  ...      14.52      19.29      14.44      -1.62   
1  5569.47  5550.44  ...    -581.91    -984.09   -1230.89   -1600.45   
2   362.95   207.27  ...      17.82     -51.66     -48.29     -59.99   
3  -746.50  -709.53  ...     122.34      93.03      93.03      68.81   
4   -45.20    -5.04  ...     -37.87     -61.85     -27.15     -21.18   

   FLUX.3192  FLUX.3193  FLUX.3194  FLUX.3195  FLUX.3196  FLUX.3197  
0      13.33      45.50      31.93      35.78     269.43      57.72  
1   -1824.53   -2061.17   -2265.98   -2366.19   -2294.86   -2034.72  
2     -82.10    -174.54     -95.23    -162.68     -36.79      30.63  
3       9.81      20.75      20.25    -120.81    -257.56    -215.41  
4     -33.76     -85.34     -81.46     -61.98     -69.34     -17.84  

[5 rows x 3198 columns]

The two DataFrames have exactly the same type of data.

**Q: Why do we have a training dataset and a test dataset?**

Ans: We create machine learning models using the training dataset to train a computer so that it can learn from that data. It, thus, can make predictions based on what it has learnt. While learning, the computer tries to find a pattern in the data, how one information affects another, which information is the most critical one, etc.

The test dataset is used to test the accuracy of the model that you have built. The higher the accuracy, the higher is the prediction capability of the machine.

After creating the DataFrames, it is a good practice to find out the number of rows and columns that exist in a DataFrame. You can do this exercise by using the `shape` keyword.

In [ ]:
# Student Action: Find the number of rows and columns in the 'exo_train_df' DataFrame.
exoTrain.shape

(5087, 3198)

So, there are 5087 rows and 3198 columns in the `exo_train_df` DataFrame.

In [ ]:
# Student Action: How many rows and columns are there in the 'exo_test_df'?
exoTest.shape

(570, 3198)

There are 570 rows and 3198 columns in the `exo_test_df` DataFrame.

---

#### Activity 2: Check For The Missing Values^

In most of the cases, we do not get complete datasets. They either have some values missing from the rows and columns or they do not have standardized values.

For example: If there is a date column in a dataset, then there is a huge chance that some of the dates are entered in the `DD-MM-YYYY` format, some in the `MM-DD-YYYY` format and so on.

So, before going ahead with the analysis, it is a good idea to check whether the dataset has any missing values.

In [ ]:
# Teacher Action: Check for the missing values using the 'isnull()' function.
exoTrain.isnull()

LABEL  FLUX.1  FLUX.2  FLUX.3  FLUX.4  FLUX.5  FLUX.6  FLUX.7  FLUX.8  \
0     False   False   False   False   False   False   False   False   False   
1     False   False   False   False   False   False   False   False   False   
2     False   False   False   False   False   False   False   False   False   
3     False   False   False   False   False   False   False   False   False   
4     False   False   False   False   False   False   False   False   False   
...     ...     ...     ...     ...     ...     ...     ...     ...     ...   
5082  False   False   False   False   False   False   False   False   False   
5083  False   False   False   False   False   False   False   False   False   
5084  False   False   False   False   False   False   False   False   False   
5085  False   False   False   False   False   False   False   False   False   
5086  False   False   False   False   False   False   False   False   False   

      FLUX.9  ...  FLUX.3188  FLUX.3189  FLUX.3190  FLUX.3191  FLUX.3192  \
0      False  ...      False      False      False      False      False   
1      False  ...      False      False      False      False      False   
2      False  ...      False      False      False      False      False   
3      False  ...      False      False      False      False      False   
4      False  ...      False      False      False      False      False   
...      ...  ...        ...        ...        ...        ...        ...   
5082   False  ...      False      False      False      False      False   
5083   False  ...      False      False      False      False      False   
5084   False  ...      False      False      False      False      False   
5085   False  ...      False      False      False      False      False   
5086   False  ...      False      False      False      False      False   

      FLUX.3193  FLUX.3194  FLUX.3195  FLUX.3196  FLUX.3197  
0         False      False      False      False      False  
1         False      False      False      False      False  
2         False      False      False      False      False  
3         False      False      False      False      False  
4         False      False      False      False      False  
...         ...        ...        ...        ...        ...  
5082      False      False      False      False      False  
5083      False      False      False      False      False  
5084      False      False      False      False      False  
5085      False      False      False      False      False  
5086      False      False      False      False      False  

[5087 rows x 3198 columns]

To check for missing values in a DataFrame, use `isnull()` function. If a DataFrame has a missing value, then the `isnull()` function will return `True` else it will return `False`.

As you can see, the `isnull()` function has returned the DataFrame with a lot of `False` values as an output. There are $5087\times3198=16268226$ values in the DataFrame. It is not feasible to check so many values manually. So, we need a better approach to check for missing values.

We can call the `sum()` function on the `exo_train_df.isnull()` statement. It will return the sum of `True` values for every column in a DataFrame. If a column does not have any missing values, then it will return `0` else a number greater than `0`.

In [ ]:
# Teacher Action: Use the 'sum()' function to find the total number of True values in each column.
exoTrain.isnull().sum()

LABEL        0
FLUX.1       0
FLUX.2       0
FLUX.3       0
FLUX.4       0
            ..
FLUX.3193    0
FLUX.3194    0
FLUX.3195    0
FLUX.3196    0
FLUX.3197    0
Length: 3198, dtype: int64

We can see that a lot of columns have `0` missing values. But still, we cannot manually see whether all the columns have missing values or not because the list of columns is too long to be seen in this notebook. There are `3198` columns to search. If there were very few columns, then we would not need to go any further to check for the missing values.

The `columns` keyword returns an array of all the columns in a DataFrame. To get a column from a DataFrame just, write the name of the column inside the square brackets in the single or double inverted quote after writing the name of the DataFrame.
For example: If you want to get the values of `FLUX.1` column from the `exo_train_df`, then write `exo_train_df['FLUX.1']`.

In [ ]:
# Teacher Action: View all the columns in the 'exo_train_df' DataFrame.
exoTrain.columns

Index(['LABEL', 'FLUX.1', 'FLUX.2', 'FLUX.3', 'FLUX.4', 'FLUX.5', 'FLUX.6',
       'FLUX.7', 'FLUX.8', 'FLUX.9',
       ...
       'FLUX.3188', 'FLUX.3189', 'FLUX.3190', 'FLUX.3191', 'FLUX.3192',
       'FLUX.3193', 'FLUX.3194', 'FLUX.3195', 'FLUX.3196', 'FLUX.3197'],
      dtype='object', length=3198)


We, again, need a better approach. We will create a variable called `num_missing_values` to store the total number of values that are missing. Then, we will iterate through each column and within each column, we will iterate through each item to check for the missing values. If the `isnull()` function for a column returns `True`, then we will increase the value of the `num_missing_values` by `1` else we will not do anything.

In [ ]:
# Teacher Action: Get the values of the 'FLUX.1' column from a DataFrame.
exoTrain['FLUX.1']

0         93.85
1        -38.88
2        532.64
3        326.52
4      -1107.21
         ...   
5082     -91.91
5083     989.75
5084     273.39
5085       3.82
5086     323.28
Name: FLUX.1, Length: 5087, dtype: float64

In [ ]:
# Student Action: Get all the values of the 'FLUX.2' column.
exoTrain['FLUX.2']

0         83.81
1        -33.83
2        535.92
3        347.39
4      -1112.59
         ...   
5082     -92.97
5083     891.01
5084     278.00
5085       2.09
5086     306.36
Name: FLUX.2, Length: 5087, dtype: float64

Now, using the `columns` keyword and the square brackets method of getting all items in a column, we will check for the missing values in the entire DataFrame.

In [ ]:
# Teacher Action: Iterate through the 'exo_train_df' DataFrame to find the total number of missing values.
nullCount = 0
for column in exoTrain.columns:
  for item in exoTrain[column].isnull():
    if item == True:
      nullCount += 1
print(nullCount)
# Here, we have created the num_missing_values which will store all the number of missing values in the DataFrame.

# Now, we will iterate through every column in the DataFrame, then will iterate through every item in each column.


0


As you can see, there are no missing values in the DataFrame because the final value of the `num_missing_values` is `0`.

Now let's find the number of non missing values by replacing `True` with `False` in the above code and store it in variable `non_missing_values`.

In [ ]:
# Student Action: In the above code replace 'True' with 'False' and get the number of non missing values.
# The output should be 16268226.
fillCount = 0
for column in exoTrain.columns:
  for item in exoTrain[column].isnull():
    if item == False:
      fillCount += 1
print(fillCount)

16268226


In [ ]:
exoTrain.shape[0] * exoTrain.shape[1]

16268226

As we can see, the output is 16,268,226. It is the sum of all the values which are False. That means there are no missing values because the total number of values in the `exo_train_df` is 16,268,226 which is exactly the same as the total number of non-missing values.

---

#### Activity 3: Slicing A DataFrame Using The `iloc[]` Function^^^

 We want to plot the scatter plots and line plots for 6 stars. For each of these stars we will create a Pandas series which will have the brightness levels starting from `FLUX.1` to `FLUX.3197`.

 Effectively, we need to create 6 Pandas series.

Let's create a Pandas series for the first star in the `exo_train_df`. Let's store the series in a variable called `star_0`. To do this, we need to use the `iloc` function.

In [ ]:
# Teacher Action: Create a Pandas series from a Pandas DataFrame using the 'iloc[]' function.
star_0 = exoTrain.iloc[0,:]

Inside, the `iloc[]` function, the digit `0` indicates the first row (located at index `0`) of the `exo_train_df` DataFrame and the colon (`:`) symbol denotes *collect all the values from the first column till the last column*, i.e., all the columns starting from `LABEL` to `FLUX.3197` columns.

**Syntax:**

`dataframe_name.iloc[row_position_start : row_position_end, column_position_start : column_position_end]`

In this syntax:

- `row_position_start` denotes the position of the row in the DataFrame **starting** from whose values you want to take in the new Pandas series or DataFrame.
- `row_position_end` denotes the position of the row in the DataFrame till whose values you want to take in the new Pandas series or DataFrame.
- `column_position_start` denotes the position of the column in the DataFrame **starting** from whose values you want to take in the new Pandas series or DataFrame.
- `column_position_end` denotes the position of the column in the DataFrame till whose values you want to take in the new Pandas series or DataFrame.

You can verify manually whether we have extracted the values from the first row or not by viewing the first 5 rows of the DataFrame using the `head()` function.

In [ ]:
# Student Action: Compare the values of the 'star_0' Pandas series with the first row in the 'exo_train_df' DataFrame.
exoTrain.head()

LABEL   FLUX.1   FLUX.2   FLUX.3   FLUX.4   FLUX.5   FLUX.6  FLUX.7  \
0      2    93.85    83.81    20.10   -26.98   -39.56  -124.71 -135.18   
1      2   -38.88   -33.83   -58.54   -40.09   -79.31   -72.81  -86.55   
2      2   532.64   535.92   513.73   496.92   456.45   466.00  464.50   
3      2   326.52   347.39   302.35   298.13   317.74   312.70  322.33   
4      2 -1107.21 -1112.59 -1118.95 -1095.10 -1057.55 -1034.48 -998.34   

    FLUX.8  FLUX.9  ...  FLUX.3188  FLUX.3189  FLUX.3190  FLUX.3191  \
0   -96.27  -79.89  ...     -78.07    -102.15    -102.15      25.13   
1   -85.33  -83.97  ...      -3.28     -32.21     -32.21     -24.89   
2   486.39  436.56  ...     -71.69      13.31      13.31     -29.89   
3   311.31  312.42  ...       5.71      -3.73      -3.73      30.05   
4 -1022.71 -989.57  ...    -594.37    -401.66    -401.66    -357.24   

   FLUX.3192  FLUX.3193  FLUX.3194  FLUX.3195  FLUX.3196  FLUX.3197  
0      48.57      92.54      39.32      61.42       5.08     -39.54  
1      -4.86       0.76     -11.70       6.46      16.00      19.93  
2     -20.88       5.06     -11.80     -28.91     -70.02     -96.67  
3      20.03     -12.67      -8.77     -17.31     -17.35      13.98  
4    -443.76    -438.54    -399.71    -384.65    -411.79    -510.54  

[5 rows x 3198 columns]

We have confirmed that we indeed created a Pandas series using the `iloc[]` function.

In [ ]:
# Student Action: Verify whether the series created is actually a Pandas series using the 'type()' function.
type(star_0)

pandas.core.series.Series

In [ ]:
type(exoTrain)

pandas.core.frame.DataFrame

As you can see, `star_0` is a Pandas series.

In [ ]:
# Student Action: Using the 'iloc[]' function, create a Pandas series for the second star and store it in a variable called 'star_1'.
# Use the 'head()' function to display only the first 5 values of the series.
star_1 = exoTrain.iloc[1,:]
star_1.head()

LABEL      2.00
FLUX.1   -38.88
FLUX.2   -33.83
FLUX.3   -58.54
FLUX.4   -40.09
Name: 1, dtype: float64

In [ ]:
# Student Action: Using the 'iloc[]' function, create a Pandas series for the third star and store it in a variable called 'star_2'.
star_2 = exoTrain.iloc[2,:]
star_2.head()

LABEL       2.00
FLUX.1    532.64
FLUX.2    535.92
FLUX.3    513.73
FLUX.4    496.92
Name: 2, dtype: float64

We have created a Pandas series for each of the first three stars. Now, let's create the same for each of the last three stars in the DataFrame.

In [ ]:
# Student Action: Display the last 5 rows of the 'exo_train_df' DataFrame using the 'tail()' function.
exoTrain.tail()

LABEL  FLUX.1  FLUX.2  FLUX.3  FLUX.4  FLUX.5  FLUX.6  FLUX.7  FLUX.8  \
5082      1  -91.91  -92.97  -78.76  -97.33  -68.00  -68.24  -75.48  -49.25   
5083      1  989.75  891.01  908.53  851.83  755.11  615.78  595.77  458.87   
5084      1  273.39  278.00  261.73  236.99  280.73  264.90  252.92  254.88   
5085      1    3.82    2.09   -3.29   -2.88    1.66   -0.75    3.85   -0.03   
5086      1  323.28  306.36  293.16  287.67  249.89  218.30  188.86  178.93   

      FLUX.9  ...  FLUX.3188  FLUX.3189  FLUX.3190  FLUX.3191  FLUX.3192  \
5082  -30.92  ...     139.95     147.26     156.95     155.64     156.36   
5083  492.84  ...     -26.50      -4.84     -76.30     -37.84    -153.83   
5084  237.60  ...     -26.82     -53.89     -48.71      30.99      15.96   
5085    3.28  ...      10.86      -3.23      -5.10      -4.61      -9.82   
5086  118.93  ...      71.19       0.97      55.20      -1.63      -5.50   

      FLUX.3193  FLUX.3194  FLUX.3195  FLUX.3196  FLUX.3197  
5082     151.75     -24.45     -17.00       3.23      19.28  
5083    -136.16      38.03     100.28     -45.64      35.58  
5084      -3.47      65.73      88.42      79.07      79.43  
5085      -1.50      -4.65     -14.55      -6.41      -2.55  
5086     -25.33     -41.31     -16.72     -14.09      27.82  

[5 rows x 3198 columns]

In [ ]:
# Student Action: Using the 'iloc[]' function, create a Pandas series for the last star and store it in a variable called 'star_5086'.
star_5086 = exoTrain.iloc[5086,:]
star_5086.head()

LABEL       1.00
FLUX.1    323.28
FLUX.2    306.36
FLUX.3    293.16
FLUX.4    287.67
Name: 5086, dtype: float64

In [ ]:
# Student Action: Using the 'iloc[]' function, create a Pandas series for the second-last star and store it in a variable called 'star_5085'.
star_5085 = exoTrain.iloc[5085,:]

In [ ]:
# Student Action: Using the 'iloc[]' function, create a Pandas series for the third-last star and store it in a variable called 'star_5084'.
star_5084 = exoTrain.iloc[5084,:]

So, in this class we learnt:

1. The transit method of detecting exoplanets

2. How to import data from a CSV file and create a Pandas DataFrame from the file.

3. How to check for the missing values.

4. How to slice a DataFrame using the `iloc[]` function.

In the next class, we will learn how to create a scatter plot and a line plot to visualise data.

---